mava/systems/q_learning/anakin/rec_qmix.py

```diff
@@ -0,0 +1,697 @@
+
+    # Initialise env steps and training steps
+    t0_act = jnp.zeros((cfg.arch.n_devices, cfg.system.update_batch_size), dtype=int)
+    t0_train = jnp.zeros((cfg.arch.n_devices, cfg.system.update_batch_size), dtype=int)
+
+    # Keys passed to learner
+    first_keys = jax.random.split(key, (cfg.arch.n_devices * cfg.system.update_batch_size))
+    first_keys = first_keys.reshape((cfg.arch.n_devices, cfg.system.update_batch_size, -1))
+
+    # Initial learner state.
+    learner_state = LearnerState(
+        first_obs,
+        first_term,
+        first_term_or_trunc,
+        init_hidden_state,
+        env_state,
+        t0_act,
+        t0_train,
+        opt_state,
+        buffer_state,
+        params,
+        first_keys,
+    )
+
+    return (env, eval_env), q_net, q_mixer, opt, rb, learner_state, logger, key
+
+
+def make_update_fns(
+    cfg: DictConfig,
+    env: MarlEnv,
+    q_net: RecQNetwork,
+    mixer: QMixNetwork,
+    opt: optax.GradientTransformation,
+    rb: TrajectoryBuffer,
+) -> Any:
+    def select_eps_greedy_action(
+        action_selection_state: ActionSelectionState,
+        obs: Observation,
+        term_or_trunc: Array,
+    ) -> Tuple[ActionSelectionState, Array]:
+        """Select action to take in eps-greedy way. Batch and agent dims are included."""
+
+        params, hidden_state, t, key = action_selection_state
+
+        eps = jnp.maximum(
+            cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
+        )
+
+        obs = tree.map(lambda x: x[jnp.newaxis, ...], obs)
+        term_or_trunc = tree.map(lambda x: x[jnp.newaxis, ...], term_or_trunc)
+
+        next_hidden_state, eps_greedy_dist = q_net.apply(
+            params, hidden_state, (obs, term_or_trunc), eps
+        )
+
+        # Random key splitting
+        new_key, explore_key = jax.random.split(key, 2)
+
+        action = eps_greedy_dist.sample(seed=explore_key)
+        action = action[0, ...]  # (1, B, A) -> (B, A)
+
+        # repack new selection params
+        next_action_selection_state = ActionSelectionState(
+            params, next_hidden_state, t + cfg.arch.num_envs, new_key
+        )
+        return next_action_selection_state, action
+
+    def action_step(action_state: ActionState, _: Any) -> Tuple[ActionState, Dict]:
+        """Selects an action, steps global env, stores timesteps in global rb and repacks the
+        parameters for the next step.
+        """
+
+        action_selection_state, env_state, buffer_state, obs, terminal, term_or_trunc = action_state
+
+        next_action_selection_state, action = select_eps_greedy_action(
+            action_selection_state, obs, term_or_trunc
+        )
+
+        next_env_state, next_timestep = jax.vmap(env.step)(env_state, action)
+
+        # Get reward
+        reward = jnp.mean(
+            next_timestep.reward, axis=-1, keepdims=True
+        )  # NOTE: Combine agent rewards, since QMIX is cooperative.
```

Comment:
nit
```suggestion
        # NOTE: Combine agent rewards, since QMIX is cooperative.
        reward = jnp.mean(next_timestep.reward, axis=-1, keepdims=True)
```